http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectHandler.java

```diff
@@ -0,0 +1,782 @@
+        ChannelFuture channelFuture = clientBootstrap.connect();
+        registerListener(channelHandlerContext, channelFuture, httpCarbonRequest, httpRequest);
+    }
+
+    /**
+     * Register channel future listener on channel future.
+     *
+     * @param channelHandlerContext Channel handler context
+     * @param channelFuture         Chanel future
+     * @param httpCarbonRequest     Carbon request
+     * @param httpRequest           http request
+     */
+    private void registerListener(ChannelHandlerContext channelHandlerContext, ChannelFuture channelFuture,
+            HTTPCarbonMessage httpCarbonRequest, HttpRequest httpRequest) {
+        channelFuture.addListener(new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture future) throws Exception {
+                if (future.isSuccess() && future.isDone()) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Connected to the new channel " + future.channel().id() + " and getting ready to "
+                                + "write request.");
+                    }
+                    long channelStartTime = channelHandlerContext.channel().attr(Constants.ORIGINAL_CHANNEL_START_TIME)
+                            .get();
+                    int timeoutOfOriginalRequest = channelHandlerContext.channel()
+                            .attr(Constants.ORIGINAL_CHANNEL_TIMEOUT).get();
+                    setChannelAttributes(channelHandlerContext, future, httpCarbonRequest, channelStartTime,
+                            timeoutOfOriginalRequest);
+                    long remainingTimeForRedirection = getRemainingTimeForRedirection(channelStartTime,
+                            timeoutOfOriginalRequest);
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Remaining time for redirection is : " + remainingTimeForRedirection);
+                    }
+                    future.channel().pipeline().addBefore(Constants.REDIRECT_HANDLER, Constants.IDLE_STATE_HANDLER,
+                            new IdleStateHandler(remainingTimeForRedirection, remainingTimeForRedirection, 0,
+                                    TimeUnit.MILLISECONDS));
+                    future.channel().write(httpRequest);
+                    future.channel().writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT);
+                    /* if the previous channel is not original channel, closes it after sending the request through
+                     new channel*/
+                    if (channelHandlerContext != originalChannelContext) {
+                        channelHandlerContext.close();
+                    }
+                } else {
+                    LOG.error("Error occurred while trying to connect to redirect channel.", future.cause());
+                    exceptionCaught(channelHandlerContext, future.cause());
```

Comment:
Are we logging the exception twice?

Reply:
This is fixed now. https://github.com/wso2/carbon-transports/pull/415